### PR TITLE
Start migration of domain-like functions located in Facade to Domain

### PR DESCRIPTION
**Pull request by RossStewart**
Switch to a new method to get all permissions (not privileges) for a user. These have either a groupUri or a granteeUri, so we'd know which are from a group. We want to move some functions that rely on domain logic within the Facade to Domain, this is the beginning of that process.

### CODE REVIEW

File: src/Domain/Services/PermissionService.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            IRepository<Group, int> groupRepository,
+            IRepository<Permission, int> permissionRepository)
+        {
+            this.groupRepository = groupRepository;
+            this.permissionRepository = permissionRepository;
+        }
+
+        public IEnumerable<Permission> GetAllPermissionsForUser(string who)
+        {
+            if (who == string.Empty)
```
Comment (lewisrenfrew) [inline src/Domain/Services/PermissionService.cs:27]: <code>string.IsNullOrEmpty(who)</code> might be safer

File: src/Domain/Services/PrivilegeService.cs [hunk 2]
```diff
@@ -24,9 +25,9 @@ public class PrivilegeService : IPrivilegeService
 
         public IEnumerable<Privilege> GetPrivileges(string who)
         {
-            if (who == String.Empty)
+            if (who == string.Empty)
```
Comment (lewisrenfrew) [inline src/Domain/Services/PrivilegeService.cs:28]: same again

File: src/Resources/Resources.csproj [hunk 3]
```diff
@@ -6,6 +6,11 @@
     <RootNamespace>Linn.Authorisation.Resources</RootNamespace>
   </PropertyGroup>
 
+  <ItemGroup>
+    <None Remove="IndividualPermissionRequestResource" />
+    <None Remove="NewFile1.txt" />
+  </ItemGroup>
```
Comment (lewisrenfrew) [inline src/Resources/Resources.csproj:12]: Not sure what these lines are doing

File: tests/Integration/Service.Tests/PermissionsModuleSpecs/WhenGettingPermissionsForAUser.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+            var p = new GroupPermission(new Group("group", true), new Privilege("Name"), "/employees/33087");
+            var permissions = new List<Permission>
+                                            { 
+                                                new IndividualPermission("/employees/33087", new Privilege(privilegeName), "/employees/7004"),
+                                                new IndividualPermission("/employees/33087", new Privilege(privilegeName), "/employees/7004"),
+                                                new GroupPermission(new Group("adminz", true), new Privilege(privilegeName), "/employees/7004")
+                                            };
+            
+            this.PermissionService.GetAllPermissionsForUser(Arg.Any<string>())
```
Comment (lewisrenfrew) [inline tests/Integration/Service.Tests/PermissionsModuleSpecs/WhenGettingPermissionsForAUser.cs:34]: blank line

File: tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    using NUnit.Framework;
+    using Permissions;
+
+    public class WhenGettingPermissionsForAUser : ContextBase
+    {
+        private readonly string privilegeName = "do.admin.stuuuff";
+        private readonly string privilegeName2 = "do-someother-stuuuff";
+        private readonly string privilegeName3 = "do-hings";
+
+
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs:20]: blank lines. (StyleCope should catch these too I think)

File: tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            {
+                if (p is IndividualPermission)
+                {
+                    individualPermissions.Add((IndividualPermission)p);
+                }
+                else
+                {
+                    groupPermissions.Add((GroupPermission)p);
+                }
+            }
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs:56]: Not really sure what's going on here. Why are we adding permissions to these lists in this test body?

File: tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+            result.ToList().Count.Should().Be(3);
+            groupPermissions.Should().Contain(x => x.Privilege.Name == this.privilegeName3 && x.GranteeGroup.Name == "adminz");
+            individualPermissions.Should().Contain(x => x.Privilege.Name == this.privilegeName && x.GranteeUri == "/employees/133");
+            individualPermissions.Should().Contain(x => x.Privilege.Name == this.privilegeName2 && x.GranteeUri == "/employees/3006");
+        }
+        [Test]
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs:63]: StyleCop will want a blank line between method bodies

File: tests/Unit/Domain.Tests/PrivilegeServiceTests/WhenGettingPrivilegesForUser.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            this.GroupRepository.FindAll().Returns(groups.AsQueryable());
+
+            var privOnlyInGroup = new Privilege { Name = "be-cool.admin", Id = 4, Active = true };
+
+            var groupPermissionsToReturn = new List<Permission>
+                                           {
+                                               new GroupPermission(group1, privOnlyInGroup, "/employees/7004")
+                                           };
+
+            this.PermissionRepository.FilterBy(Arg.Any<Expression<Func<Permission, bool>>>())
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.Tests/PrivilegeServiceTests/WhenGettingPrivilegesForUser.cs:54]: blank line

File: tests/Unit/Facade.Tests/PermissionServiceTests/WhenGettingPermissionsForAUser.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    using Linn.Authorisation.Domain.Groups;
+    using Linn.Authorisation.Domain.Permissions;
+    using Linn.Common.Facade;
+    using NSubstitute;
+    using NUnit.Framework;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Linq.Expressions;
+
+    using Microsoft.EntityFrameworkCore;
```
Comment (lewisrenfrew) [inline tests/Unit/Facade.Tests/PermissionServiceTests/WhenGettingPermissionsForAUser.cs:17]: unused reference? I'd be suprised if the tests needed a reference to EntityFramework. Correct me if I'm wrong though!

File: tests/Unit/Facade.Tests/PermissionServiceTests/WhenGettingPermissionsForAUser.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        {
+
+            var permissions = new List<Permission>
+                                            {
+                                                new IndividualPermission("/employees/133", new Privilege(this.privilegeName), "/employees/7004"),
+                                                new IndividualPermission("/employees/3006", new Privilege(this.privilegeName2), "/employees/7004"),
+                                                new GroupPermission(new Group("adminz", true), new Privilege(this.privilegeName3), "/employees/7004")
+                                            };
+
+            this.PermissionService.GetAllPermissionsForUser(Arg.Any<string>()).Returns(permissions);
```
Comment (lewisrenfrew) [inline tests/Unit/Facade.Tests/PermissionServiceTests/WhenGettingPermissionsForAUser.cs:38]: blank

File: tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs (hunk 6 above)
Comment (RossStewart) [reply tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs:56]: Permissions assigned to both groups and individual members can both have the same name. We add to two separate lists for this test to check that we are returning the right type of permission type each time if that makes sense?

Comment (RossStewart) [reply tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs:56]: Yeah, that's totally fair. I based this off of an existing Privileges test as they're quite similar so I can always make changes to that too so it's a bit easier to read.

**Comment by richardiphillips** [comment]
> Just responded to a comment there. I don't think my Stylecop has been on? Either that or it's not recognising a few rules/the need for new lines at the end of files (Iain and I noticed this post commit on Friday). Just been making the changes manually and about to submit.

It won't be on by default. You have to add the stylecop extension within resharper. I had planned to show you that before we got onto coding but you were too fast 🛩️

File: tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs (hunk 6 above)
Comment (lewisrenfrew) [reply tests/Unit/Domain.Tests/PermissionsTests/WhenGettingPermissionsForUser.cs:56]: You can just cast the list entry to the right type inside your Should.Contain() to see the fields specific to that type

File: src/Service.Host/client/src/components/Privileges.js [hunk 11]
```diff
@@ -127,9 +125,9 @@ function ViewPrivileges({
     };
     const dispatchcreatePrivilege = () => createPrivilege(initialName);
 
-    const changeUser = e => {
-        selectUser(e.target.value);
-    };
+    // const changeUser = e => {
```
Comment (iaincrawf) [inline src/Service.Host/client/src/components/Privileges.js:128]: Would just delete this wee function if we're not using it, and can probably take out the use effect and call get all privileges in the container on load or something along those lines

**Review by iaincrawf — APPROVED** [review]

File: src/Service.Host/client/src/components/Root.js [hunk 12]
```diff
@@ -69,6 +70,27 @@ const Root = ({ store }) => (
...
+                                    exact
+                                    path="/authorisation/viewpermissions"
+                                    render={() => <Permissions store={store} />}
+                                    store={store}
+                                />
+
+                                <Route
+                                    exact
+                                    path="/authorisation/permissions/:id"
+                                    render={() => <Permissions store={store} />}
```
[the remark this answers is not in the document]
Comment (iaincrawf) [reply src/Service.Host/client/src/components/Root.js:91]: Actually should likely only need one of the new routes above too depending which you've used, either /viewpermissions/ or /permissions/

**Pull request merged**